File: diagnostic-capacitor/ios/Sources/DiagnosticPlugin/DiagnosticPlugin.swift
import Foundation
import Capacitor

@objc(DiagnosticPlugin)
public class DiagnosticPlugin: CAPPlugin {

    private lazy var location = LocationModule()
    private lazy var bluetooth = BluetoothModule(plugin: self)
    private lazy var camera = CameraModule()
    private lazy var notifications = NotificationsModule(plugin: self)
    private lazy var wifi = WifiModule(plugin: self)
    private lazy var microphone = MicrophoneModule(plugin: self)
    private lazy var motion = MotionModule(plugin: self)
    private lazy var reminders = RemindersModule(plugin: self)
    private lazy var calendar = CalendarModule()
    private lazy var contacts = ContactsModule()

    // -----------------------
    // Location
    // -----------------------

    @objc func getLocationAuthorizationStatus(_ call: CAPPluginCall) { location.getLocationAuthorizationStatus(call) }
    @objc func requestLocationAuthorization(_ call: CAPPluginCall) { location.requestLocationAuthorization(call) }
    @objc func openLocationSettings(_ call: CAPPluginCall) { location.openLocationSettings(call) }
    @objc func switchToLocationSettings(_ call: CAPPluginCall) { location.switchToLocationSettings(call) }
    @objc func isLocationEnabled(_ call: CAPPluginCall) { location.isLocationEnabled(call) }
    @objc func isLocationAvailable(_ call: CAPPluginCall) { location.isLocationAvailable(call) }
    @objc func isGpsLocationEnabled(_ call: CAPPluginCall) { location.isGpsLocationEnabled(call) }
    @objc func isNetworkLocationEnabled(_ call: CAPPluginCall) { location.isNetworkLocationEnabled(call) }
    @objc func isGpsLocationAvailable(_ call: CAPPluginCall) { location.isGpsLocationAvailable(call) }
    @objc func isNetworkLocationAvailable(_ call: CAPPluginCall) { location.isNetworkLocationAvailable(call) }
    @objc func getLocationMode(_ call: CAPPluginCall) { location.getLocationMode(call) }
    @objc func isCompassAvailable(_ call: CAPPluginCall) { location.isCompassAvailable(call) }
    @objc func isLocationAuthorized(_ call: CAPPluginCall) { location.isLocationAuthorized(call) }
    @objc func getLocationAccuracyAuthorization(_ call: CAPPluginCall) { location.getLocationAccuracyAuthorization(call) }
    @objc func requestTemporaryFullAccuracyAuthorization(_ call: CAPPluginCall) { location.requestTemporaryFullAccuracyAuthorization(call) }

    // -----------------------
    // Bluetooth
    // -----------------------

    @objc func switchToBluetoothSettings(_ call: CAPPluginCall) { bluetooth.switchToBluetoothSettings(call) }

    @objc func isBluetoothAvailable(_ call: CAPPluginCall) { bluetooth.isBluetoothAvailable(call) }
    @objc func isBluetoothEnabled(_ call: CAPPluginCall) { bluetooth.isBluetoothEnabled(call) }
    @objc func hasBluetoothSupport(_ call: CAPPluginCall) { bluetooth.hasBluetoothSupport(call) }
    @objc func hasBluetoothLESupport(_ call: CAPPluginCall) { bluetooth.hasBluetoothLESupport(call) }
    @objc func hasBluetoothLEPeripheralSupport(_ call: CAPPluginCall) { bluetooth.hasBluetoothLEPeripheralSupport(call) }
    @objc func setBluetoothState(_ call: CAPPluginCall) { bluetooth.setBluetoothState(call) }
    @objc func getBluetoothState(_ call: CAPPluginCall) { bluetooth.getBluetoothState(call) }
    @objc func getBluetoothAuthorizationStatuses(_ call: CAPPluginCall) { bluetooth.getBluetoothAuthorizationStatuses(call) }
    @objc func requestBluetoothAuthorization(_ call: CAPPluginCall) { bluetooth.requestBluetoothAuthorization(call) }
    @objc func ensureBluetoothManager(_ call: CAPPluginCall) { bluetooth.ensureBluetoothManager(call) }
    @objc func getBluetoothAuthorizationStatus(_ call: CAPPluginCall) { bluetooth.getBluetoothAuthorizationStatus(call) }

    // -----------------------
    // Camera
    // -----------------------

    @objc func isCameraPresent(_ call: CAPPluginCall) { camera.isCameraPresent(call) }
    @objc func requestCameraAuthorization(_ call: CAPPluginCall) { camera.requestCameraAuthorization(call) }
    @objc func getCameraAuthorizationStatus(_ call: CAPPluginCall) { camera.getCameraAuthorizationStatus(call) }
    @objc func getCameraAuthorizationStatuses(_ call: CAPPluginCall) { camera.getCameraAuthorizationStatuses(call) }

    // -----------------------
    // Notifications
    // -----------------------

    @objc func isRemoteNotificationsEnabled(_ call: CAPPluginCall) { notifications.isRemoteNotificationsEnabled(call) }
    @objc func getRemoteNotificationTypes(_ call: CAPPluginCall) { notifications.getRemoteNotificationTypes(call) }
    @objc func isRegisteredForRemoteNotifications(_ call: CAPPluginCall) { notifications.isRegisteredForRemoteNotifications(call) }
    @objc func getRemoteNotificationsAuthorizationStatus(_ call: CAPPluginCall) { notifications.getRemoteNotificationsAuthorizationStatus(call) }
    @objc func requestRemoteNotificationsAuthorization(_ call: CAPPluginCall) { notifications.requestRemoteNotificationsAuthorization(call) }
    @objc func switchToNotificationSettings(_ call: CAPPluginCall) { notifications.switchToNotificationSettings(call) }

    // -----------------------
    // Wifi
    // -----------------------

    @objc func isWifiAvailable(_ call: CAPPluginCall) { wifi.isWifiAvailable(call) }
    @objc func isWifiEnabled(_ call: CAPPluginCall) { wifi.isWifiEnabled(call) }
    @objc func requestLocalNetworkAuthorization(_ call: CAPPluginCall) { wifi.requestLocalNetworkAuthorization(call) }
    @objc func getLocalNetworkAuthorizationStatus(_ call: CAPPluginCall) { wifi.getLocalNetworkAuthorizationStatus(call) }

    // -----------------------
    // Microphone
    // -----------------------


    @objc func isMicrophoneAuthorized(_ call: CAPPluginCall) {microphone.isMicrophoneAuthorized(call)}
    @objc func getMicrophoneAuthorizationStatus(_ call: CAPPluginCall) {microphone.getMicrophoneAuthorizationStatus(call)}
    @objc func requestMicrophoneAuthorization(_ call: CAPPluginCall) {microphone.requestMicrophoneAuthorization(call)}

    // -----------------------
    // Motion
    // -----------------------

    @objc func isMotionAvailable(_ call: CAPPluginCall) {motion.isMotionAvailable(call)}
    @objc func isMotionRequestOutcomeAvailable(_ call: CAPPluginCall) {motion.isMotionRequestOutcomeAvailable(call)}
    @objc func getMotionAuthorizationStatus(_ call: CAPPluginCall) {motion.getMotionAuthorizationStatus(call)}
    @objc func requestMotionAuthorization(_ call: CAPPluginCall) {motion.requestMotionAuthorization(call)}

    // -----------------------
    // Reminders
    // -----------------------

    @objc func getRemindersAuthorizationStatus(_ call: CAPPluginCall) {reminders.getRemindersAuthorizationStatus(call)}
    @objc func isRemindersAuthorized(_ call: CAPPluginCall) {reminders.isRemindersAuthorized(call)}
    @objc func requestRemindersAuthorization(_ call: CAPPluginCall) {reminders.requestRemindersAuthorization(call)}

    // -----------------------
    // Calendar
    // -----------------------
    
    @objc func getCalendarAuthorizationStatus(_ call: CAPPluginCall) { calendar.getCalendarAuthorizationStatus(call) }
    @objc func isCalendarAuthorized(_ call: CAPPluginCall) { calendar.isCalendarAuthorized(call) }
    @objc func requestCalendarAuthorization(_ call: CAPPluginCall) { calendar.requestCalendarAuthorization(call) }

    // -----------------------
    // Contacts
    // -----------------------

    @objc func getAddressBookAuthorizationStatus(_ call: CAPPluginCall) { contacts.getAddressBookAuthorizationStatus(call) }
    @objc func isAddressBookAuthorized(_ call: CAPPluginCall) { contacts.isAddressBookAuthorized(call) }
    @objc func requestAddressBookAuthorization(_ call: CAPPluginCall) { contacts.requestAddressBookAuthorization(call) }
}

File: diagnostic-capacitor/android/src/main/java/com/noesis/diagnostic/DiagnosticPlugin.java
package com.noesis.diagnostic;

import android.Manifest;
import android.os.Build;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

import com.noesis.diagnostic.modules.BluetoothModule;
import com.noesis.diagnostic.modules.CameraModule;
import com.noesis.diagnostic.modules.LocationModule;
import com.noesis.diagnostic.modules.NotificationsModule;
import com.noesis.diagnostic.modules.WifiModule;

@CapacitorPlugin(
    name = "DiagnosticPlugin",
    permissions = {
        @Permission(
            alias = "location",
            strings = {
                Manifest.permission.ACCESS_COARSE_LOCATION,
                Manifest.permission.ACCESS_FINE_LOCATION
            }
        ),
        @Permission(
            alias = "backgroundLocation",
            strings = {
                Manifest.permission.ACCESS_BACKGROUND_LOCATION
            }
        ),
        @Permission(
            alias = "bluetooth",
            strings = {
                Manifest.permission.BLUETOOTH_ADVERTISE,
                Manifest.permission.BLUETOOTH_CONNECT,
                Manifest.permission.BLUETOOTH_SCAN
            }
        ),
        @Permission(
            alias = "camera",
            strings = {
                Manifest.permission.CAMERA
            }
        ),
        @Permission(
            alias = "cameraStorageLegacy",
            strings = {
                Manifest.permission.READ_EXTERNAL_STORAGE,
                Manifest.permission.WRITE_EXTERNAL_STORAGE
            }
        ),
        @Permission(
            alias = "cameraStorage33",
            strings = {
                Manifest.permission.READ_MEDIA_IMAGES,
                Manifest.permission.READ_MEDIA_VIDEO
            }
        ),
        @Permission(
            alias = "cameraStorage34",
            strings = {
                Manifest.permission.READ_MEDIA_IMAGES,
                Manifest.permission.READ_MEDIA_VIDEO,
                Manifest.permission.READ_MEDIA_VISUAL_USER_SELECTED
            }
        ),
        @Permission(
            alias = "notifications",
            strings = {
                Manifest.permission.POST_NOTIFICATIONS
            }
        )
    }
)
public class DiagnosticPlugin extends Plugin implements BluetoothModule.BluetoothEventEmitter {

    private LocationModule location;
    private BluetoothModule bluetooth;
    private CameraModule cameraModule;
    private NotificationsModule notifications;
    private WifiModule wifi;

     @Override
    public void load() {
        super.load();

        location = new LocationModule(this);
        bluetooth = new BluetoothModule(this, this);
        cameraModule = new CameraModule(getContext());
        notifications = new NotificationsModule(this);
        wifi = new WifiModule(this);

        bluetooth.load();
    }

    @Override
    protected void handleOnDestroy() {
        if (bluetooth != null) {
            bluetooth.handleOnDestroy();
        }
        super.handleOnDestroy();
    }

    @Override
    public void emitBluetoothStateChange(String state) {
        JSObject data = new JSObject();
        data.put("state", state);
        notifyListeners("bluetoothStateChange", data);
    }

    // -----------------------
    // Location
    // -----------------------

    @PluginMethod
    public void requestLocationAuthorization(PluginCall call) {
        location.setLocationEverAsked();
        call.getData().put("mode", call.getString("mode", "when_in_use"));
        requestPermissionForAlias("location", call, "onLocationPermissionResult");
    }

    @PluginMethod
    public void getLocationAuthorizationStatus(PluginCall call) {
        location.getLocationAuthorizationStatus(call);
    }

    @PluginMethod
    public void isLocationAvailable(PluginCall call) {
        location.isLocationAvailable(call);
    }

    @PluginMethod
    public void isLocationEnabled(PluginCall call) {
        location.isLocationEnabled(call);
    }

    @PluginMethod
    public void openLocationSettings(PluginCall call) {
        location.openLocationSettings(call);
    }

    @PluginMethod
    public void getLocationMode(PluginCall call) {
        location.getLocationMode(call);
    }

    @PluginMethod
    public void isGpsLocationEnabled(PluginCall call) {
        location.isGpsLocationEnabled(call);
    }

    @PluginMethod
    public void isNetworkLocationEnabled(PluginCall call) {
        location.isNetworkLocationEnabled(call);
    }

    @PluginMethod
    public void isGpsLocationAvailable(PluginCall call) {
        location.isGpsLocationAvailable(call);
    }

    @PluginMethod
    public void isNetworkLocationAvailable(PluginCall call) {
        location.isNetworkLocationAvailable(call);
    }

    @PluginMethod
    public void switchToLocationSettings(PluginCall call) {
        location.switchToLocationSettings(call);
    }

    @PluginMethod
    public void isCompassAvailable(PluginCall call) {
        location.isCompassAvailable(call);
    }

    @PluginMethod
    public void isLocationAuthorized(PluginCall call) {
        location.isLocationAuthorized(call);
    }

    @PluginMethod
    public void getLocationAccuracyAuthorization(PluginCall call) {
        location.getLocationAccuracyAuthorization(call);
    }

    @PluginMethod
    public void requestTemporaryFullAccuracyAuthorization(PluginCall call) {
        location.requestTemporaryFullAccuracyAuthorization(call);
    }

    @PermissionCallback
    private void onLocationPermissionResult(PluginCall call) {
        location.onLocationPermissionResult(call);

        String mode = call.getData().optString("mode", "when_in_use");
        if ("always".equalsIgnoreCase(mode) && Build.VERSION.SDK_INT == Build.VERSION_CODES.Q) {
            requestPermissionForAlias("backgroundLocation", call, "onBackgroundLocationPermissionResult");
        }
    }

    @PermissionCallback
    private void onBackgroundLocationPermissionResult(PluginCall call) {
        location.onBackgroundLocationPermissionResult(call);
    }

    // -----------------------
    // Bluetooth
    // -----------------------

    @PluginMethod
    public void switchToBluetoothSettings(PluginCall call) {
        bluetooth.switchToBluetoothSettings(call);
    }

    @PluginMethod
    public void isBluetoothAvailable(PluginCall call) {
        bluetooth.isBluetoothAvailable(call);
    }

    @PluginMethod
    public void isBluetoothEnabled(PluginCall call) {
        bluetooth.isBluetoothEnabled(call);
    }

    @PluginMethod
    public void hasBluetoothSupport(PluginCall call) {
        bluetooth.hasBluetoothSupport(call);
    }

    @PluginMethod
    public void hasBluetoothLESupport(PluginCall call) {
        bluetooth.hasBluetoothLESupport(call);
    }

    @PluginMethod
    public void hasBluetoothLEPeripheralSupport(PluginCall call) {
        bluetooth.hasBluetoothLEPeripheralSupport(call);
    }

    @PluginMethod
    public void setBluetoothState(PluginCall call) {
        bluetooth.setBluetoothState(call);
    }

    @PluginMethod
    public void getBluetoothState(PluginCall call) {
        bluetooth.getBluetoothState(call);
    }

    @PluginMethod
    public void getBluetoothAuthorizationStatuses(PluginCall call) {
        bluetooth.getBluetoothAuthorizationStatuses(call);
    }

    @PluginMethod
    public void requestBluetoothAuthorization(PluginCall call) {
        bluetooth.requestBluetoothAuthorization(call);
    }

    @PluginMethod
    public void ensureBluetoothManager(PluginCall call) {
        bluetooth.ensureBluetoothManager(call);
    }

    @PluginMethod
    public void getBluetoothAuthorizationStatus(PluginCall call) {
        bluetooth.getBluetoothAuthorizationStatus(call);
    }

    @PermissionCallback
    private void onBluetoothPermissionResult(PluginCall call) {
        bluetooth.onBluetoothPermissionResult(call);
    }

    public void requestBluetoothPermissions(PluginCall call) {
        requestPermissionForAlias("bluetooth", call, "onBluetoothPermissionResult");
    }

    // -----------------------
    // Camera
    // -----------------------

    @PluginMethod
    public void isCameraPresent(PluginCall call) {
        JSObject result = new JSObject();
        result.put("present", cameraModule.isCameraPresent());
        call.resolve(result);
    }

    @PluginMethod
    public void getCameraAuthorizationStatuses(PluginCall call) {
        boolean storage = call.getBoolean("storage", false);
        JSObject result = new JSObject();
        result.put("statuses", cameraModule.getCameraAuthorizationStatuses(storage, getActivity()));
        call.resolve(result);
    }

    @PluginMethod
    public void getCameraAuthorizationStatus(PluginCall call) {
        boolean storage = call.getBoolean("storage", false);
        JSObject result = new JSObject();
        result.put("status", cameraModule.getCameraAuthorizationStatus(storage, getActivity()));
        call.resolve(result);
    }

    @PluginMethod
    public void requestCameraAuthorization(PluginCall call) {
        boolean storage = call.getBoolean("storage", false);

        call.getData().put("storage", storage);
        cameraModule.markPermissionsRequested(cameraModule.getPermissions(storage));

        requestPermissionForAlias("camera", call, "onCameraPermissionResult");
    }

    @PermissionCallback
    private void onCameraPermissionResult(PluginCall call) {
        boolean storage = call.getData().optBoolean("storage", false);

        if (!storage) {
            JSObject result = new JSObject();
            result.put("status", cameraModule.getCameraAuthorizationStatus(false, getActivity()));
            call.resolve(result);
            return;
        }

        requestPermissionForAlias(getCameraStorageAlias(), call, "onCameraStoragePermissionResult");
    }

    @PermissionCallback
    private void onCameraStoragePermissionResult(PluginCall call) {
        boolean storage = call.getData().optBoolean("storage", false);

        JSObject result = new JSObject();
        result.put("status", cameraModule.getCameraAuthorizationStatus(storage, getActivity()));
        call.resolve(result);
    }

    private String getCameraStorageAlias() {
        if (Build.VERSION.SDK_INT >= 34) {
            return "cameraStorage34";
        } else if (Build.VERSION.SDK_INT >= 33) {
            return "cameraStorage33";
        } else {
            return "cameraStorageLegacy";
        }
    }

    // -----------------------
    // Notifications
    // -----------------------

    @PluginMethod
    public void isRemoteNotificationsEnabled(PluginCall call) {
        notifications.isRemoteNotificationsEnabled(call);
    }

    @PluginMethod
    public void getRemoteNotificationTypes(PluginCall call) {
        notifications.getRemoteNotificationTypes(call);
    }

    @PluginMethod
    public void isRegisteredForRemoteNotifications(PluginCall call) {
        notifications.isRegisteredForRemoteNotifications(call);
    }

    @PluginMethod
    public void getRemoteNotificationsAuthorizationStatus(PluginCall call) {
        notifications.getRemoteNotificationsAuthorizationStatus(call);
    }

    @PluginMethod
    public void requestRemoteNotificationsAuthorization(PluginCall call) {
        notifications.requestRemoteNotificationsAuthorization(call);
    }

    @PluginMethod
    public void switchToNotificationSettings(PluginCall call) {
        notifications.switchToNotificationSettings(call);
    }

    @PermissionCallback
    private void onNotificationsPermissionResult(PluginCall call) {
        notifications.onNotificationsPermissionResult(call);
    }

    public void requestNotificationsPermission(PluginCall call) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            requestPermissionForAlias("notifications", call, "onNotificationsPermissionResult");
        } else {
            notifications.onNotificationsPermissionNotRequired(call);
        }
    }
    
    // -----------------------
    // Wifi
    // -----------------------

    @PluginMethod
    public void switchToWifiSettings(PluginCall call) {
        wifi.switchToWifiSettings(call);
    }

    @PluginMethod
    public void isWifiAvailable(PluginCall call) {
        wifi.isWifiAvailable(call);
    }

    @PluginMethod
    public void isWifiEnabled(PluginCall call) {
        wifi.isWifiEnabled(call);
    }

    @PluginMethod
    public void setWifiState(PluginCall call) {
        wifi.setWifiState(call);
    }
}